File: snippets/rusty.py
from mpi4py import MPI
import numpy as np

from rusty_tree.distributed import DistributedTree

# Setup communicator
comm = MPI.COMM_WORLD

# Cartesian points at this process
points = np.random.rand(100000, 3)

# Generate a balanced tree
balanced = DistributedTree.from_global_points(points, True, comm)

# Generate an unbalanced tree
unbalanced = DistributedTree.from_global_points(points, False, comm)

# Trees map between points and the keys that they correspond to contain
point = balanced.points[0].head # grab a point
print(balanced[point]) # find the key that it maps

# Trees map between keys and the points they contain
key = balanced.keys[0].head # grab a point
print(balanced[key]) # find the points that this key contains

# Trees implement iterator protocol as well as slicing and indexing
# (without copy of underlying Rust data)
# Slice of 10 keys
key_slice = balanced.keys[:10]

# Slice of 10 points
point_slice = balanced.points[:10]

def foo(key):
    """Some key processor function"""
    pass

# Iterating in this way avoids copying data from Rust to Python
for key in key_slice:
    foo(key)

# Copy only performed when printing in Python
print(point_slice)
print(key_slice)

File: snippets/rusty.rs
use rand::prelude::*;
use rand::SeedableRng;

use mpi::{
    environment::Universe,
    topology::UserCommunicator,
    traits::*
};

use rusty_tree::{
    distributed::DistributedTree,
};


fn main () {
    // Generate a set of randomly distributed points
    let mut range = StdRng::seed_from_u64(0);
    let between = rand::distributions::Uniform::from(0.0..1.0);
    let mut points: Vec<[PointType; 3]> = Vec::new();
    let npoints = 1000000;

    for _ in 0..npoints {
        points.push([
            between.sample(&mut range),
            between.sample(&mut range),
            between.sample(&mut range),
        ])
    }

    // Setup an MPI environment
    let universe: Universe = mpi::initialize().unwrap();
    let world: UserCommunicator = universe.world();
    let comm = world.duplicate();

    // Unbalanced
    let unbalanced: DistributedTree = DistributedTree::new(&points, false, &comm)

    // Balanced
    let balanced_tree: DistributedTree = DistributedTree::new(&points, true, &comm)
 
}
